### ISSUE TITLE
`requires-python = "~=3.10"` uses docker container python 3.11 and not a managed python 3.10

### ISSUE THREAD

**Issue by gabriel-vanzandycke** [astral-sh/uv#7682]
I use a docker container bundled with python 3.11.9 but my app requires python ~=3.10. 
Running `uv sync` uses python system rather than the expected python version `~=3.10`.
When enforcing `==3.10`, uv downloads and uses a managed python version as expected.

**Dockerfile**
```dockerfile
FROM python:3.11.9-slim-bookworm
COPY --from=ghcr.io/astral-sh/uv:0.4.10 /uv /bin/uv
ENTRYPOINT /bin/bash
ADD . /root

RUN --mount=type=cache,target=/root/.cache/uv \
    --mount=type=bind,source=uv.lock,target=uv.lock \
    --mount=type=bind,source=pyproject.toml,target=pyproject.toml \
    uv sync --frozen --no-install-project --no-install-workspace > /tmp/uv_sync.log 2>&1
```

**pyprojects.toml**
```pyproject.toml
[project]
name = "app"
version = "1.6.0"
description = ""
requires-python = "~=3.10"
dependencies = [
    "jupyter>=1.0",
]

[tool.uv]
package = true
```

**With `requires-python = "~=3.10"` in pyproject.toml:**
Inspecting the beginning of uv sync output shows it is using python 3.11.9.
```
$ docker build -t app:latest -f Dockerfile .
$ docker run -it --rm app:latest
# head -1 /tmp/uv_sync.log
Using Python 3.11.9 interpreter at: /usr/local/bin/python3
```

**With `requires-python = "==3.10"` in pyproject.toml:**
When requiring `3.10` instead, inspecting the uv sync output shows it is using python 3.10 as expected.
```
$ docker build -t app:latest -f Dockerfile .
$ docker run -it --rm app:latest
root@d93a4e3c8011:/# head -1 /tmp/uv_sync.log
Using Python 3.10.0
```

**Comment by gabriel-vanzandycke** [comment]
Thanks. I didn't know that.

**Comment by gabriel-vanzandycke** [comment]
I still have the problem with `~=3.10.0` though

**Comment by gabriel-vanzandycke** [comment]
it's not easy to replicate my exact setup in a minimum working example, but when I use `~=3.10.0`, running
```
uv run jupyter notebook
```
and importing one of my dependency within a notebook fails with "module not found". It all seems like uv run is not loading/using the environment created previously.

**Comment by charliermarsh** [comment]
If `~=3.10.0` is giving you `3.11.X`, that may be a bug.

**Comment by charliermarsh** [comment]
I cannot produce that:
```
Using Python 3.11.7
error: The requested interpreter resolved to Python 3.11.7, which is incompatible with the project's Python requirement: `>=3.10.0, <3.11`
```

**Comment by charliermarsh** [comment]
Happy to reopen if we're able to find a reproduction.
